File: packages/sdk-ts/src/resources/chains.ts
import type { Result } from "../errors.js";
import { retryingRequest } from "../http.js";

// Wave 2 (Phase A coverage). Mirrors the chain-runs REST surface mounted by
// apps/api/src/routes/chains.ts:
//   POST /api/v1/chain-runs          — create + start a chain run
//   GET  /api/v1/chain-runs/:id      — chain run + steps
//   GET  /api/v1/reviews/:id/chain   — chain context for a review
//
// The shapes here re-state @gatewerk/shared's ChainDefinition / ChainRun /
// ChainStep types. We cannot import @gatewerk/shared from the SDK because the
// SDK ships standalone to npm and the consumer must not be forced to install
// the workspace `shared` package. The shapes are kept structurally compatible
// — see packages/shared/src/api/schemas/chains.ts for the source-of-truth
// Zod schemas (ChainDefinitionSchema, ChainRunObjectSchema, ChainStepObject-
// Schema). Any drift here is a typing-quality bug; the wire format is what
// the backend Zod schema emits.

export interface ChainAssigneeUser {
  kind: "user";
  email?: string;
  user_id?: string;
}

export interface ChainAssigneeRole {
  kind: "role";
  role: "admin" | "reviewer";
}

export interface ChainAssigneeExternalToken {
  kind: "external_token";
  expires_in_seconds?: number;
  grace_period_seconds?: number;
  note?: string;
}

export type ChainAssigneeSpec =
  | ChainAssigneeUser
  | ChainAssigneeRole
  | ChainAssigneeExternalToken;

export type ChainStepRejectionPolicy = "abort" | "continue" | "branch";

export interface ChainDefinitionStep {
  id: string;
  name?: string;
  description?: string;
  template: string;
  assignee: ChainAssigneeSpec;
  timeout_seconds?: number;
  priority?: string;
  metadata?: Record<string, unknown>;
  depends_on?: string[];
  rejection_policy?: ChainStepRejectionPolicy;
  rejection_branch_to?: number;
}

export type ChainRejectionPolicy = "terminate" | "back_one" | "restart";
export type ChainMode = "sequential" | "parallel" | "mixed";

export interface ChainDefinition {
  version: "1.0";
  name?: string;
  description?: string;
  mode: ChainMode;
  rejection_policy?: ChainRejectionPolicy;
  metadata?: Record<string, unknown>;
  steps: ChainDefinitionStep[];
}

export interface ChainCreateInput {
  definition: ChainDefinition;
  initial_payload: Record<string, unknown>;
  callback_url?: string;
  metadata?: Record<string, unknown>;
}

export type ChainStepStatus =
  | "pending"
  | "active"
  | "approved"
  | "rejected"
  | "expired"
  | "skipped"
  | "superseded";

export interface ChainStepObject {
  object?: "chain_step";
  id: string;
  chain_run_id: string;
  step_number: number;
  review_id: string | null;
  assignee_spec: Record<string, unknown>;
  depends_on: string[] | null;
  status: ChainStepStatus;
  materialized_at: string | null;
  rejection_policy?: ChainStepRejectionPolicy | null;
  rejection_branch_to?: number | null;
}

export type ChainRunStatus = "active" | "completed" | "rejected" | "aborted";

export interface ChainRunObject {
  object?: "chain_run";
  id: string;
  project_id: string;
  template_id: string | null;
  name: string | null;
  mode: string;
  rejection_policy: string;
  status: ChainRunStatus;
  metadata: Record<string, unknown> | null;
  created_by: string;
  created_at: string;
  completed_at: string | null;
  steps?: ChainStepObject[];
  // Only present on POST response — convenience pointer to the first step's
  // review id so callers don't have to scan steps[] (matches routes/chains.ts).
  step_1_review_id?: string;
  // Only present on GET /reviews/:id/chain — the caller's step pointer.
  current_step_number?: number | null;
}

// All three endpoints return the chain object wrapped in the standard
// envelope `{ object: "chain_run", ...fields }` (apps/api/src/routes/chains.ts
// uses envelope("chain_run", …)). Callers receive the wrapped shape directly.
export class ChainsResource {
  constructor(
    private baseUrl: string,
    private headers: () => Record<string, string>,
  ) {}

  // Retry helper centralised in ../http.ts.
  private async request<T>(url: string, init?: RequestInit): Promise<Result<T>> {
    return retryingRequest<T>(url, init, this.headers);
  }

  async create(input: ChainCreateInput): Promise<Result<ChainRunObject>> {
    return this.request(`${this.baseUrl}/api/v1/chain-runs`, {
      method: "POST",
      body: JSON.stringify(input),
    });
  }

  async get(runId: string): Promise<Result<ChainRunObject>> {
    return this.request(`${this.baseUrl}/api/v1/chain-runs/${runId}`, { method: "GET" });
  }

  async getForReview(reviewId: string): Promise<Result<ChainRunObject>> {
    return this.request(`${this.baseUrl}/api/v1/reviews/${reviewId}/chain`, { method: "GET" });
  }
}

File: packages/sdk-py/gatewerk/resources/chains.py
"""Chains resource for the Gatewerk Python SDK.

Mirrors the chain-runs REST surface mounted by apps/api/src/routes/chains.ts:

    POST /api/v1/chain-runs           — create + start a chain run
    GET  /api/v1/chain-runs/:id       — chain run + steps
    GET  /api/v1/reviews/:id/chain    — chain context for a review

The wire shapes (ChainDefinition, ChainRunObject, ChainStepObject) live in
``gatewerk.types`` — they restate
``packages/shared/src/api/schemas/chains.ts`` Zod schemas as Pydantic
models with ``extra="allow"`` for forward-compat. The SDK can't import
``@gatewerk/shared`` at runtime, so any drift between this file and the
backend Zod schema is a typing-quality bug; the wire format is what the
backend Zod emits.
"""

from __future__ import annotations

from typing import Any, Optional, Union

from .._base import BaseResource
from ..types import ChainDefinition, ChainRunObject


class ChainsResource(BaseResource):

    def create(
        self,
        definition: Union[ChainDefinition, dict[str, Any]],
        initial_payload: dict[str, Any],
        *,
        callback_url: Optional[str] = None,
        metadata: Optional[dict[str, Any]] = None,
        timeout: Optional[float] = None,
    ) -> ChainRunObject:
        """Create + start a chain run.

        Args:
            definition: Chain definition (a :class:`ChainDefinition` or
                an equivalent dict). Defines the steps, mode, and rejection
                policy.
            initial_payload: Starting payload for the first step.
            callback_url: Optional webhook URL to call when chain status
                changes.
            metadata: Optional run-scoped metadata.

        Returns:
            :class:`ChainRunObject`: The created run with all materialized
            steps. ``step_1_review_id`` is populated for convenience.
        """
        body: dict[str, Any] = {
            "definition": (
                definition.model_dump(exclude_none=True)
                if isinstance(definition, ChainDefinition)
                else definition
            ),
            "initial_payload": initial_payload,
        }
        if callback_url is not None:
            body["callback_url"] = callback_url
        if metadata is not None:
            body["metadata"] = metadata

        data = self._request("POST", "/api/v1/chain-runs", json=body, timeout=timeout)
        return ChainRunObject.model_validate(data)

    def get(self, run_id: str, *, timeout: Optional[float] = None) -> ChainRunObject:
        """Fetch the full chain run + its steps."""
        data = self._request("GET", f"/api/v1/chain-runs/{run_id}", timeout=timeout)
        return ChainRunObject.model_validate(data)

    def get_for_review(
        self, review_id: str, *, timeout: Optional[float] = None
    ) -> ChainRunObject:
        """Fetch the chain context (run + steps + current_step_number) for a review.

        Useful when you have a review id and want to know whether it's part
        of a chain and which step it's at. Returns the full
        :class:`ChainRunObject` with ``current_step_number`` populated.
        """
        data = self._request(
            "GET", f"/api/v1/reviews/{review_id}/chain", timeout=timeout
        )
        return ChainRunObject.model_validate(data)
